### PR TITLE
Add SetExpressionCompiler for reporting v2alpha

### PR DESCRIPTION
**Pull request by riemanli**
Difference between v1 and v2:
* Treat the v2 one as a pure algorithm library. 
    * Remove proto dependency.
    * The conversion from `ReportingSet.SetExpression` to `SetExpressionCompiler.SetOperationExpression` is now done by the caller. 
* The reporting set in `setOperationExpression` has to be indexed from 0 to `numReportingSets - 1`.

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)

**Comment by riemanli** [comment]
Current dependencies on/for this PR:
* main
  * **PR #951** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/951" target="_blank">Graphite</a> 
    * **PR #963** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/963" target="_blank">Graphite</a> 
      * **PR #959** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/959" target="_blank">Graphite</a> 
        * **PR #968** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/968" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/968?utm_source=stack-comment).

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @chenweiw and @riemanli)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt` line 107 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968#-NTteAF4ER1fwaiz0bM-:-NTteAF4ER1fwaiz0bM0:bfssna8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c2c41fc749eb353a881d3f50f08465f97b3e4ee1/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt#L107)):*
> ```Kotlin
>   private val vennDiagramDecompositionCache: VennDiagramDecompositionCache = mutableMapOf()
> 
>   // For unit test only.
> ```

This is a code smell. Do we really need to verify the caching behavior? If it's not observable from the public API then it probably doesn't need to be tested. If the only observable trait is performance then we can pick that up in a performance test.
___
*[`src/test/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompilerTest.kt` line 118 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968#-NTtdpPMBcz7WyvlM0F_:-NTtdpPMBcz7WyvlM0Fa:b-faaohe) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/c2c41fc749eb353a881d3f50f08465f97b3e4ee1/src/test/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompilerTest.kt#L118)):*
> ```Kotlin
>     }
> 
>     Truth.assertThat(resultAllUnionButOne)
> ```

nit: import assertThat



_Code quote:_
```Kotlin
Truth.assertThat
```

**Review by SanjayVas — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @chenweiw)

File: src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+ * setOf(1(=b’001’), 3(=b’011’), 5(=b’101’), 7(=b’111’)).
+ */
+private typealias VennDiagramDecomposition = List<Set<PrimitiveRegion>>
+
+/**
+ * A memory cache that stores the decompositions of Venn Diagrams given different number of
+ * reporting sets.
+ */
+private typealias VennDiagramDecompositionCache =
+  MutableMap<NumberReportingSets, VennDiagramDecomposition>
```
Comment (chenweiw) [inline src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt:82]: I remember we did not have this memory in the alpha version. Is it a new feature we add in w2alpha?

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: all files reviewed, 1 unresolved discussion (waiting on @chenweiw)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt` line 82 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968#gh-1178224574:-NTyib6V23NOvOEJOzPZ:b-gpm1nh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1d32ec7f893aaa27631866248ff51097f6b6bd67/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt#L82)):*
<details><summary><i>Previously, chenweiw wrote…</i></summary><blockquote>

I remember we did not have this memory in the alpha version. Is it a new feature we add in w2alpha?
</blockquote></details>

Yes! I figured that which primitive regions form a reporting set is static and only depending on the number of reporting sets. We can save this 2^(N-1) computation by using a memory cache.

**Review by chenweiw — COMMENTED** [review]
The algorithm of calculating SetOperation expressions is correct.

**Comment by chenweiw** [comment]
> > _[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)_ status: all files reviewed, 1 unresolved discussion (waiting on @chenweiw)
> 
> _[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt` line 82 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968#gh-1178224574:-NTyib6V23NOvOEJOzPZ:b-gpm1nh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/1d32ec7f893aaa27631866248ff51097f6b6bd67/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt#L82)):_
> 
> Previously, chenweiw wrote…
> Yes! I figured that which primitive regions form a reporting set is static and only depending on the number of reporting sets. We can save this 2^(N-1) computation by using a memory cache.

It makes sense. Thank you Rieman.

**Review by chenweiw — APPROVED** [review]
The algorithm implementation of calculating the SetOperation expressions is correct.

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: all files reviewed, 1 unresolved discussion (waiting on @chenweiw)

**Review by riemanli — COMMENTED** [review]
> Dismissed @chenweiw from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @riemanli)

**Comment by riemanli** [comment]
@riemanli queued this pull request to merge with [Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/968).

**Comment by riemanli** [comment]
[Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/968) rebased this pull request as part of a merge.

**Comment by riemanli** [comment]
The [Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/968) merge of this pull request was cancelled.

**Review by riemanli — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 2 of 2 files at r2, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/968)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @riemanli)

**Pull request merged**